### PR TITLE
Support cross-tenant virtual network linking for Private DNS zones

### LINKED ISSUE #9851: Private DNS link vnet from different tenant
## Describe the bug
I have two tenants (_tenant1_, _tenant2_). Each tenant has a subscription (_sub1_, _sub2_), resource group (_rg1_, _rg2_), vnets (_vnet1_, _vnet2_). I am able to create a peering between _vnet1_ and _vnet2_ using the IAM roles configured (added user from _tenant1_ to _rg2_ with Network Contributor role)
I am trying to register _vnet2_ as a resolution network (old nomenclature) to the private zone created in _sub1_. Tried both previous and current preview version of Private DNS zone feature - both failing with similar messages.
Documentation only specifies permissions necessary to create vnet peering between different tenants - Network Contributor. Do I need any other permissions?

Is it possible at all?

`az account list` for _user1_:
``` json
[
  {
    "cloudName": "AzureCloud",
    "id": "_sub1_",
    "isDefault": true,
    "name": "Visual Studio Enterprise",
    "state": "Enabled",
    "tenantId": "_tenant1_",
    "user": {
      "name": "_user1_",
      "type": "user"
    }
  },
  {
    "cloudName": "AzureCloud",
    "id": "_sub2_",
    "isDefault": false,
    "name": "_sub2-name_",
    "state": "Enabled",
    "tenantId": "_tenant2_",
    "user": {
      "name": "_user1_",
      "type": "user"
    }
  }
]
```

**Command Name**
`az network private-dns link vnet create`

**Errors:**
```
The client has permission to perform action 'Microsoft.Network/virtualNetworks/join/action' on scope '/subscriptions/_sub1_/resourceGroups/_rg1_/providers/Microsoft.Network/privateDnsZones/_zonename_/virtualNetworkLinks/_vnet1_', however the current tenant '_tenant1_' is not authorized to access linked subscription '_sub2_'.
```

## To Reproduce:
az network private-dns link vnet create --resource-group _rg1_ --name _linkname_ --zone-name _zonename_ --virtual-network /subscriptions/_sub2_/resourceGroups/_rg2_/providers/Microsoft.Network/virtualNetworks/_vnet2_ --registration-enabled false --subscription _sub1_

### PR DESCRIPTION
**Pull request by dijyotir**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

Fixes #9851

### CODE REVIEW

**Comment by dijyotir** [comment]
@zikalino This is the same change that I did for the extension in this [PR](https://github.com/Azure/azure-cli-extensions/pull/946). Please review and approve.

**Comment by Juliehzl** [comment]
CC @MyronFanQiu, who is domain expert for Network module.

**Comment by dijyotir** [comment]
Hi @MyronFanQiu, please add this to the sprint 75.

File: src/azure-cli/azure/cli/command_modules/privatedns/tests/latest/test_privatedns_commands.py [hunk 1]
```diff
@@ -722,12 +722,10 @@ def test_PatchRecordSet_SoaRecord_ExpectRecordSetUpdated(self, resource_group):
         self.kwargs['serialNumber'] = 5
         self.kwargs['host'] = 'clitest.hostmaster.com'
         self.cmd('az network private-dns record-set soa update -g {rg} -z {zone} \
-            -e {email} -x {expireTime} -t {host} -m {minimumTtl} -f {refreshTime} -r {retryTime} -s {serialNumber}', checks=[
+            -e {email} -x {expireTime} -m {minimumTtl} -f {refreshTime} -r {retryTime} -s {serialNumber}', checks=[
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/privatedns/tests/latest/test_privatedns_commands.py]: Is there any reason for you to remove `-t`?

Comment (dijyotir) [reply src/azure-cli/azure/cli/command_modules/privatedns/tests/latest/test_privatedns_commands.py]: Yes I just rerun the test. The `-t` is used to change hostname but we don't allow to update the hostname except in certain cases. So can't really test it with a random hostname. So removed it from the test.

**Comment by MyronFanQiu** [comment]
@dijyotir Hello. Please rebase your branch against latest dev and modify the history. Then I will merge this PR.

**Comment by manvkaur** [comment]
Hi Team,

Can you please share the reference document or azure cli command to support cross tenant vnet linking for Private DNS Zone using two different service principals?

**Comment by jsntcy** [comment]
@msyyc, could you help on this?

**Comment by jiasli** [comment]
@manvkaur, this is definitely something that should be placed in the public doc.

You may take this script as a reference for "**support cross-tenant VNET linking for Private DNS Zone using two different service principals**".

We need to create 2 apps:

- `appA`
   - `tenantA` - `spA`  - create VNET
- `appB`
   - `tenantB` - `spB`  - create private-DNS zone and private-dns link
   - `tenantA` - `spBA` - write VNET (`appB` must be a guest in `tenantA`, and have write permission on the VNET)


### Log in with a user account

```sh
# Prepare args
tenantA='ca97...'
subA="414a..."
tenantB='5482...'
subB="0b1f..."

# Login with a user account which is a tenant Application Administrator
az login
```

### Create appA under tenantA
```sh
# Switch to subA under tenantA
az account set --subscription $subA

# Create appA and spA
spA=$(az ad sp create-for-rbac --role Contributor)

# Save the appId and password for later login
appIdA=$(echo $spA | jq --raw-output .appId)
passwordA=$(echo $spA | jq --raw-output .password)
```

### Create appB under tenantB

```sh
# Switch to subB under tenantB
az account set --subscription $subB

# Create appB and spB
spB=$(az ad sp create-for-rbac --role Contributor)

# Save the appId and password for later login
appIdB=$(echo $spB | jq --raw-output .appId)
passwordB=$(echo $spB | jq --raw-output .password)

# Make appB multi-tenant
# --identifier-uris must be on a tenant verified domain. See https://stackoverflow.com/a/60396506/2199657
az ad app update --id $appIdB --available-to-other-tenants --identifier-uris "https://AzureSDKTeam.onmicrosoft.com/$appIdB"
```

### Make appB a guest under tenantA

```sh
# Switch to subA under tenantA
az account set --subscription $subA

# Make appB a guest in tenantA by creating spBA
az ad sp create --id $appIdB

# Grant appB Contributor role in subA
# You may narrow the scope of the role assignment by specifying --scope
az role assignment create --assignee $appIdB --role Contributor
```

### Create VNET in subA

```sh
# Log in with appA in tenantA
az login --service-principal --username $appIdA --password $passwordA --tenant $tenantA

# Create myResourceGroupA
az group create --name myResourceGroupA --location "East US"

# Create myVnetA
az network vnet create --resource-group myResourceGroupA --name myVnetA --location eastus --address-prefix 10.2.0.0/16 --subnet-name backendSubnet --subnet-prefixes 10.2.0.0/24
```

### Create private DNS zone in subB

```sh
# Log in with appB in tenantB
az login --service-principal --username $appIdB --password $passwordB --tenant $tenantB

# Create myResourceGroupB
az group create --name myResourceGroupB --location "East US"

# Create private DNS private.contoso.com
az network private-dns zone create -g myResourceGroupB -n private.contoso.com
```

### Log in with appB in tenantA (optional)

This step is optional. If not executed, `az network private-dns link vnet create` will show a warning that the credential under `tenantA` is missing. CLI will force to use the credential under `tenantB` in `tenantA`.

```sh
# Log in with appB in tenantA
az login --service-principal --username $appIdB --password $passwordB --tenant $tenantA

# Switch back to subB
az login --service-principal --username $appIdB --password $passwordB --tenant $tenantB
```

### Create private DNS link in subB

```sh
# Create private DNS link MyDNSLink
az network private-dns link vnet create -g myResourceGroupB -n MyDNSLink -z private.contoso.com -v "/subscriptions/$subA/resourceGroups/myResourceGroupA/providers/Microsoft.Network/virtualNetworks/myVnetA" -e true
```

The request will contain following headers:

- `Authorization`: `appB`'s access token for `tenantB`
- `x-ms-authorization-auxiliary`: `appB`'s access token for `tenantA`

## References

- [Quickstart: Create an Azure private DNS zone using the Azure CLI](https://docs.microsoft.com/en-us/azure/dns/private-dns-getstarted-cli)
- [`az network private-dns link vnet create`](https://docs.microsoft.com/en-us/cli/azure/network/private-dns/link/vnet?view=azure-cli-latest#az_network_private_dns_link_vnet_create)
- [Authenticate requests across tenants](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/authenticate-multi-tenant)

**Comment by jiasli** [comment]
BTW, once private DNS is migrated to Track 2, it will also be affected by https://github.com/Azure/azure-sdk-for-python/issues/8313.

**Pull request merged**